### ISSUE TITLE
[css-scroll-snap] Multiple nested scrollers and a "default" scrollIntoView()?

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#2593]
Our intention in #1708 is that scrollIntoView()'s alignment arg is honored if specified, but if it's unspecified we honor scroll-snap; only when both are unspecified do we use the "default" sIV() alignment.

What do we do when the targetted element is nested into multiple scrollable elements, tho? Each one of them needs to be scrolled to bring the element into view.  When sIV() doesn't specify a particular alignment, how should we treat all the nested scrollers? In particular, what if *some* of them have scroll-snap specified, while others don't?

Honoring scroll-snap on the scrollers that have it, and default-aligning the rest, seems like probably the best behavior, but it's also significantly more complicated for implementations, as they have to collect and track more data as they execute the scroll operation.  What should we do?

**Comment by majido** [comment]
The proposed solution seems reasonable though as pointed out it adds complexity.

This means that we are now allowing snap and align to be simultaneously honored depending on the container. This is fine but it is odd that the only possible alignment option is the default alignment. I think this is not very user friendly and provides for an odd API. 

This issue is actually caused by the original resolution which makes snap application depend on presence of alignment.

I also have a second concern with original proposal. The way it works if a user specifies an alignment we
ignore the snap position. But that means the next user scroll **will snap**. I understand that this may be the
desired behavior in some usecases but I don't feel escaping snapping and this secondary snap induced scroll is the right **default**. 

## Alternative proposal

Since the spec has not been updated and no one has implemented then I suppose we have a chance to reconsider the original resolution. Here is an alternative idea:

Introduce a `snap` property to ScrollIntoViewOptions that explicitly controls snapping behavior independent of the alignment. By default it is `true` but then it can be set to `false` to disable snapping.

```webidl
dictionary ScrollIntoViewOptions : ScrollOptions {
  ScrollLogicalPosition block = "start";
  ScrollLogicalPosition inline = "nearest";
  boolean snap = true; 
};
```

I think giving authors an explicit way to disable snapping provides a better API. Also it is better suited if in future we decide to provide a similar option for other scrolling API (e.g., `focus()`, `scrollBy()`).

So in this particular case:
- If author has explicitly requested disabling snap via `snap:false` we don't snap any ancestor container and apply the requested alignment (which can be other than default).
- Otherwise, snap for ancestors that are snap containers, and align for ancestors that are not based on the specified or the default alignment.

**Comment by tabatkins** [comment]
If you want to disable snapping, we already have an API for that - just set `scroll-snap-type: none` on the container. We don't need to add more ways to do this, I don't think; in particular, I think it's quite weird to have a scrolling operation have the *side effect* of also permanently disabling snapping for the container.

Overall I'm kinda confused tho - you introduce the proposal with the problem that an aligned sIV() won't honor the snap, but the next scroll will; but then your proposed addition just addresses the original problem in my first post, that of what to do about intermediate snap containers during an unaligned sIV().

(Also, boolean options should default to falsey values, by convention. So if we did do this, we'd want to reverse the naming/meaning of the flag.)

**Comment by tabatkins** [comment]
Ah, okay, then yeah, `{preventSnap}` in the `ScrollOptions` sounds good.

**Comment by fantasai** [comment]
I do want to point out that one of the design constraints we agreed on was that script cannot put a scroller into a scroll position that the user cannot then return to.

**Comment by tabatkins** [comment]
That is directly contradicting our other resolution that an explicit alignment directive in a scrolling method should align accordingly, overriding any scroll-snapping.

Script shouldn't put the scroller into a user-unreachable position *by default*, but when you explicitly request it, we will do so.

**Comment by fantasai** [comment]
No, the other resolution was that the explicit alignment directive would set the scroll target, but it would then be subject to any snapping constraints. If there's none that conflict, then it's fine, you get the alignment you asked for. But if you've got mandatory snapping in a way that conflicts, then it wins, just as if you manually scrolled to the same alignment the script specified and let go.

**Comment by majido** [comment]
I also read the original resolution differently as it meant that snap *only* wins when there is no alignment. 

I am fine with fantasai's [explanation above](https://github.com/w3c/csswg-drafts/issues/2593#issuecomment-385099549). It basically amounts to snap always be in effect which is consistent and easy to reason about.

I don't think we need to talk about wining at all rather snapping operates *after* all other scrolling operations so it always applies. In the case of sIV() if the alignment brings the scroll offset to
a location which is a valid snap point then the alignment remains otherwise we adjust it to honor snapping even if it means the alignment is ignored.

So effectively:
1. Apply scroll operation (user scroll or programmatic scrolls e.g., scrollTo, scrollBy, scrollIntoView, etc.)
2. Apply scroll snapping

**Comment by tabatkins** [comment]
I'm down with that, it's just not how I interpreted the previous resolutions. ^_^

So yeah, that makes this stuff easier, then. Just scroll as normal, with all the nested scrollers aligned as normal, then snap some of them as needed.

**Comment by fantasai** [comment]
Marking needs edits for applying clarifications.
